// @nrwl/workspace/src/core/normalize-nx-json.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const normalize_nx_json_1 = require("./normalize-nx-json");
describe('normalizeNxJson', () => {
    it('should expand projects array', () => {
        const result = normalize_nx_json_1.normalizeNxJson({
            npmScope: 'nrwl',
            projects: {
                demo: { tags: [] },
                ui: { tags: [] },
                util: { tags: [] },
            },
            implicitDependencies: {
                'package.json': '*',
                'whatever.json': {
                    a: {
                        b: {
                            c: ['demo'],
                            d: {
                                e: '*',
                            },
                        },
                    },
                },
            },
        });
        expect(result.implicitDependencies).toEqual({
            'package.json': ['demo', 'ui', 'util'],
            'whatever.json': {
                a: {
                    b: {
                        c: ['demo'],
                        d: {
                            e: ['demo', 'ui', 'util'],
                        },
                    },
                },
            },
        });
    });
});
