// @nrwl/workspace/src/schematics/remove/lib/check-targets.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const testing_1 = require("@angular-devkit/schematics/testing");
const ast_utils_1 = require("@nrwl/workspace/src/utils/ast-utils");
const testing_2 = require("@nrwl/workspace/testing");
const testing_3 = require("../../../utils/testing");
const check_targets_1 = require("./check-targets");
describe('checkTargets Rule', () => {
    let tree;
    let schema;
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        tree = new testing_1.UnitTestTree(schematics_1.Tree.empty());
        tree = testing_2.createEmptyWorkspace(tree);
        schema = {
            projectName: 'ng-app',
            skipFormat: false,
            forceRemove: false,
        };
        tree = (yield testing_3.callRule(ast_utils_1.updateWorkspaceInTree((workspace) => {
            return {
                version: 1,
                projects: {
                    'ng-app': {
                        projectType: 'application',
                        schematics: {},
                        root: 'apps/ng-app',
                        sourceRoot: 'apps/ng-app/src',
                        prefix: 'happyorg',
                        architect: {
                            build: {
                                builder: '@angular-devkit/build-angular:browser',
                                options: {},
                            },
                        },
                    },
                    'ng-app-e2e': {
                        root: 'apps/ng-app-e2e',
                        sourceRoot: 'apps/ng-app-e2e/src',
                        projectType: 'application',
                        architect: {
                            e2e: {
                                builder: '@nrwl/cypress:cypress',
                                options: {
                                    cypressConfig: 'apps/ng-app-e2e/cypress.json',
                                    tsConfig: 'apps/ng-app-e2e/tsconfig.e2e.json',
                                    devServerTarget: 'ng-app:serve',
                                },
                            },
                        },
                    },
                },
            };
        }), tree));
    }));
    it('should throw an error if another project targets', () => __awaiter(void 0, void 0, void 0, function* () {
        yield expect(testing_3.callRule(check_targets_1.checkTargets(schema), tree)).rejects.toThrow();
    }));
    it('should NOT throw an error if no other project targets', () => __awaiter(void 0, void 0, void 0, function* () {
        schema.projectName = 'ng-app-e2e';
        yield expect(testing_3.callRule(check_targets_1.checkTargets(schema), tree)).resolves.not.toThrow();
    }));
    it('should not error if forceRemove is true', () => __awaiter(void 0, void 0, void 0, function* () {
        schema.forceRemove = true;
        yield expect(testing_3.callRule(check_targets_1.checkTargets(schema), tree)).resolves.not.toThrow();
    }));
});
