// @nrwl/workspace/src/schematics/move/lib/update-cypress-json.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const testing_1 = require("@angular-devkit/schematics/testing");
const workspace_1 = require("@nrwl/workspace");
const testing_2 = require("@nrwl/workspace/testing");
const testing_3 = require("../../../utils/testing");
const update_cypress_json_1 = require("./update-cypress-json");
describe('updateCypressJson Rule', () => {
    let tree;
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        tree = new testing_1.UnitTestTree(schematics_1.Tree.empty());
        tree = testing_2.createEmptyWorkspace(tree);
    }));
    it('should handle cypress.json not existing', () => __awaiter(void 0, void 0, void 0, function* () {
        tree = yield testing_3.runSchematic('lib', { name: 'my-lib' }, tree);
        expect(tree.files).not.toContain('/libs/my-destination/cypress.json');
        const schema = {
            projectName: 'my-lib',
            destination: 'my-destination',
        };
        yield expect(testing_3.callRule(update_cypress_json_1.updateCypressJson(schema), tree)).resolves.not.toThrow();
    }));
    it('should update the videos and screenshots folders', () => __awaiter(void 0, void 0, void 0, function* () {
        const cypressJson = {
            fileServerFolder: '.',
            fixturesFolder: './src/fixtures',
            integrationFolder: './src/integration',
            pluginsFile: './src/plugins/index',
            supportFile: false,
            video: true,
            videosFolder: '../../dist/cypress/libs/my-lib/videos',
            screenshotsFolder: '../../dist/cypress/libs/my-lib/screenshots',
            chromeWebSecurity: false,
        };
        tree = yield testing_3.runSchematic('lib', { name: 'my-lib' }, tree);
        tree.create('/libs/my-destination/cypress.json', JSON.stringify(cypressJson));
        expect(tree.files).toContain('/libs/my-destination/cypress.json');
        const schema = {
            projectName: 'my-lib',
            destination: 'my-destination',
        };
        tree = (yield testing_3.callRule(update_cypress_json_1.updateCypressJson(schema), tree));
        expect(workspace_1.readJsonInTree(tree, '/libs/my-destination/cypress.json')).toEqual(Object.assign(Object.assign({}, cypressJson), { videosFolder: '../../dist/cypress/libs/my-destination/videos', screenshotsFolder: '../../dist/cypress/libs/my-destination/screenshots' }));
    }));
});
