// @nrwl/workspace/src/command-line/list.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const app_root_1 = require("../utils/app-root");
const output_1 = require("../utils/output");
const plugins_1 = require("../utils/plugins");
exports.list = {
    command: 'list [plugin]',
    describe: 'Lists installed plugins, capabilities of installed plugins and other available plugins.',
    builder: (yargs) => yargs.positional('plugin', {
        default: null,
        description: 'The name of an installed plugin to query',
    }),
    handler: listHandler,
};
/**
 * List available plugins or capabilities within a specific plugin
 *
 * @remarks
 *
 * Must be run within an Nx workspace
 *
 */
function listHandler(args) {
    return __awaiter(this, void 0, void 0, function* () {
        if (args.plugin) {
            plugins_1.listPluginCapabilities(args.plugin);
        }
        else {
            const corePlugins = yield plugins_1.fetchCorePlugins();
            const communityPlugins = yield plugins_1.fetchCommunityPlugins();
            const installedPlugins = plugins_1.getInstalledPluginsFromNodeModules(app_root_1.appRootPath, corePlugins, communityPlugins);
            plugins_1.listInstalledPlugins(installedPlugins);
            plugins_1.listCorePlugins(installedPlugins, corePlugins);
            plugins_1.listCommunityPlugins(installedPlugins, communityPlugins);
            output_1.output.note({
                title: `Use "nx list [plugin]" to find out more`,
            });
        }
    });
}
