// @nrwl/workspace/src/migrations/update-9-1-0/update-lint-config.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const schematics_1 = require("@angular-devkit/schematics");
const workspace_1 = require("@nrwl/workspace");
const testing_1 = require("@nrwl/workspace/testing");
const testing_2 = require("../../utils/testing");
const workspace_2 = require("@nrwl/workspace/src/utils/workspace");
describe('Update 9.1.0', () => {
    let tree;
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        tree = schematics_1.Tree.empty();
        tree = testing_1.createEmptyWorkspace(tree);
    }));
    it('should update tslint.json', () => __awaiter(void 0, void 0, void 0, function* () {
        tree = yield testing_2.callRule(schematics_1.chain([
            workspace_2.updateWorkspace((workspace) => {
                workspace.projects.add({
                    name: 'proj1',
                    root: 'proj1',
                    architect: {
                        lint: {
                            builder: '@angular-devkit/build-angular:tslint',
                        },
                    },
                });
                workspace.projects.add({
                    name: 'proj2',
                    root: 'proj2',
                    architect: {
                        lint: {
                            builder: '@angular-devkit/build-angular:tslint',
                        },
                    },
                });
            }),
            workspace_1.updateJsonInTree('proj1/tslint.json', () => ({
                rules: {},
            })),
            workspace_1.updateJsonInTree('proj2/tslint.json', () => ({
                rules: {},
                linterOptions: {
                    exclude: ['whatever'],
                },
            })),
        ]), tree);
        const result = yield testing_2.runMigration('update-lint-config-9-1-0', {}, tree);
        expect(workspace_1.readJsonInTree(result, 'proj1/tslint.json')).toEqual({
            rules: {},
            linterOptions: {
                exclude: ['!**/*'],
            },
        });
        expect(workspace_1.readJsonInTree(result, 'proj2/tslint.json')).toEqual({
            rules: {},
            linterOptions: {
                exclude: ['!**/*', 'whatever'],
            },
        });
    }));
    it('should update .eslintrc', () => __awaiter(void 0, void 0, void 0, function* () {
        tree = yield testing_2.callRule(schematics_1.chain([
            workspace_2.updateWorkspace((workspace) => {
                workspace.projects.add({
                    name: 'proj1',
                    root: 'proj1',
                    architect: {
                        lint: {
                            builder: '@nrwl/linter:lint',
                            options: {
                                config: 'proj1/.eslintrc',
                            },
                        },
                    },
                });
                workspace.projects.add({
                    name: 'proj2',
                    root: 'proj2',
                    architect: {
                        lint: {
                            builder: '@nrwl/linter:lint',
                            options: {
                                config: 'proj2/.eslintrc',
                            },
                        },
                    },
                });
            }),
            workspace_1.updateJsonInTree('proj1/.eslintrc', () => ({
                rules: {},
            })),
            workspace_1.updateJsonInTree('proj2/.eslintrc', () => ({
                rules: {},
                ignorePatterns: ['whatever'],
            })),
        ]), tree);
        const result = yield testing_2.runMigration('update-lint-config-9-1-0', {}, tree);
        expect(workspace_1.readJsonInTree(result, 'proj1/.eslintrc')).toEqual({
            rules: {},
            ignorePatterns: ['!**/*'],
        });
        expect(workspace_1.readJsonInTree(result, 'proj2/.eslintrc')).toEqual({
            rules: {},
            ignorePatterns: ['!**/*', 'whatever'],
        });
    }));
});
